### PR TITLE
Added extensibility points for searchView/context

### PR DESCRIPTION
**Pull request by digeff**
2nd attempt to merge: https://github.com/microsoft/vscode/pull/109049
This PR fixes part of #107996

I tested this by writing a mini extension that contributes a menu for the search results:
```
	"contributes": {
		"commands": [
			{
				"command": "set-breakpoints-on-search-results.helloWorld",
				"title": "Hello World"
			}
		],
		"menus": {
			"search/context": [
				{
					"command": "set-breakpoints-on-search-results.helloWorld"
				}
			]
		}
	},
```

I got the menu to appear, and I got the expected parameters when I clicked on it:
image

cc @roblourens

### CODE REVIEW

File: src/vs/workbench/api/common/extHostSearch.ts [hunk 1]
```diff
@@ -37,8 +38,27 @@ export class ExtHostSearch implements ExtHostSearchShape {
...
 		@IExtHostRpcService private extHostRpc: IExtHostRpcService,
 		@IURITransformerService protected _uriTransformer: IURITransformerService,
-		@ILogService protected _logService: ILogService
-	) { }
+		@ILogService protected _logService: ILogService,
+		commands: ExtHostCommands,
+	) {
+		commands.registerArgumentProcessor({
+			processArgument: arg => {
+				if (arg && Array.isArray(arg)) {
```
Comment (rheh) [inline src/vs/workbench/api/common/extHostSearch.ts:46]: There is a lot of indentation here.  Could we switch the logic and return early. e.g.
`if (!arg || !Array.isArray(arg)) {
   return arg;
}'

This removes the needs for the else and makes the code a little more readbale.

Same could be done to the === 13 line.

File: src/vs/workbench/api/common/extHostSearch.ts [hunk 2]
```diff
@@ -37,8 +38,27 @@ export class ExtHostSearch implements ExtHostSearchShape {
...
+					return arg.map(matchContext => {
+						if (matchContext.$mid === 13 /* SearchViewContextMid */) {
```
Comment (rheh) [inline src/vs/workbench/api/common/extHostSearch.ts:48]: Could the 13 (magic number) be moved into a named constant to reveal its meaning

File: src/vs/workbench/contrib/search/browser/searchView.ts [hunk 3]
```diff
@@ -85,6 +87,43 @@ export enum SearchViewPosition {
 	Panel
 }
 
+export type SearchViewContextMid = 13;
+export const SearchViewContextMid = 13;
```
Comment (rheh) [inline src/vs/workbench/contrib/search/browser/searchView.ts:91]: Still unsure what the 13 actually does?  Is it a order / position?

Comment (roblourens) [reply src/vs/workbench/contrib/search/browser/searchView.ts:91]: It's an ID used to tag data transfer object

File: src/vs/workbench/contrib/search/browser/searchView.ts [hunk 4]
```diff
@@ -780,13 +819,29 @@ export class SearchView extends ViewPane {
...
+		let locations: Location[] = [];
+		if (match instanceof FileMatch) {
+			locations = match.matches().map(toLocation);
+		} else if (match instanceof Match) {
+			locations = [toLocation(match)];
+		} else if (match instanceof FolderMatch) {
+			locations = flatten(match.matches().map(fileMatch => fileMatch.matches().map(toLocation)));
+		}
+
+		let context = {
```
Comment (roblourens) [inline src/vs/workbench/contrib/search/browser/searchView.ts:832]: @jrieken are you good with this? Making the context which could be a common type `{ locations: [...] }`, or we could name it `location` but give it an array. Or we could use `{ location: Location }[]`, but we have to think about multiselect in the future.

Comment (roblourens) [reply src/vs/workbench/contrib/search/browser/searchView.ts:832]: Also note that `renderableMatch` is like an "internal context" that the argument processor removes.

Comment (jrieken) [reply src/vs/workbench/contrib/search/browser/searchView.ts:832]: For multiselect the custom tree does the following: the first element is always the active tree item, the second element is all selected items. So, you select 7 elements and right click on the 3rd, then it is `[3rd element, [7 elements]]`

For the actual data type I am unsure. A location seems to be working well but not differentiating between a single match and group of matches (file) and a group of groups (folder) will likely not make everyone happy.

**Review by jrieken — CHANGES_REQUESTED** [review]
inject should be changed, the overall data type selection gives me a weird feeling because we don't look at this holistically

**Comment by roblourens** [comment]
@JacksonKearl I guess I should think about whether this could work in the search editor too. Can we contribute context menu commands to the search editor that would get the same context as those commands in a tree or are you stuck with whatever the editor gives it?

That's a complication because the search editor structures the displayed results differently than the tree and we should try to have a context that works for both.

File: src/vs/workbench/contrib/search/browser/searchView.ts (hunk 4 above)
Comment (roblourens) [reply src/vs/workbench/contrib/search/browser/searchView.ts:832]: Oh ok, I didn't know that. I think the explorer is just always a list of URIs.

Someone could reverse engineer the groupings from our flat list if they wanted to. I still think it's a workable solution but yeah it's not what I would design from scratch just for search. Besides this simple common type, we could reuse `TextSearchMatch` from the findTextInFiles proposed api. I'd rather use something that matches what's in the UI more directly though. And that won't be stabilized soon.

**Comment by JacksonKearl** [comment]
We can contribute commands to specifically search-result editor context menus, the difficulty would be in generating the arguments to pass to the context view, especially in the case of saved search results where we'd need to parse it from the text. If we don't support saved it wouldn't be too difficult, we could simply retain the last search result object.

**Comment by roblourens** [comment]
> especially in the case of saved search results where we'd need to parse it from the text

You already do that so I can navigate to the result, right? 

Although I would worry about giving a stale result to an extension command, could be unexpected. Could only have it work in live results.

**Comment by JacksonKearl** [comment]
We parse out the destination URI and line/char, but not a full `TextSearchMatch`.

**Comment by roblourens** [comment]
I think my feelings on this were that we should wait for other search API to be finalized.

**Pull request closed without merging**